js/modules/k6/experimental/fs/module.go

```diff
@@ -165,3 +170,114 @@ func (f *File) Stat() *goja.Promise {
 
 	return promise
 }
+
+// Read the file's content, and writes it into the provided Uint8Array.
+//
+// Resolves to either the number of bytes read during the operation
+// or EOF (null) if there was nothing more to read.
+//
+// It is possible for a read to successfully return with 0 bytes.
+// This does not indicate EOF.
+func (f *File) Read(into goja.Value) *goja.Promise {
+	promise, resolve, reject := promises.New(f.vu)
+
+	if common.IsNullish(into) {
+		reject(newFsError(TypeError, "read() failed; reason: into cannot be null or undefined"))
+		return promise
+	}
+
+	// We expect the into argument to be a `Uint8Array` instance
+	intoObj := into.ToObject(f.vu.Runtime())
+	uint8ArrayConstructor := f.vu.Runtime().Get("Uint8Array")
+	if isUint8Array := intoObj.Get("constructor").SameAs(uint8ArrayConstructor); !isUint8Array {
+		reject(newFsError(TypeError, "read() failed; reason: into argument must be a Uint8Array"))
+		return promise
+	}
+
+	// Obtain the underlying ArrayBuffer from the Uint8Array
+	ab, ok := intoObj.Get("buffer").Export().(goja.ArrayBuffer)
+	if !ok {
+		reject(newFsError(TypeError, "read() failed; reason: into argument cannot be interpreted as ArrayBuffer"))
+		return promise
+	}
+
+	// Obtain the underlying byte slice from the ArrayBuffer.
+	// Note that this is not a copy, and will be modified by the Read operation
+	// in place.
+	buffer := ab.Bytes()
```

Comment:
Unfortunately I think this makes the whole operation really easy to race and I will show two examples that I have tested.

Both of them are not likely code, but this code racing IMO is pretty bad and will need to be fixed before this becomes not experimental:

```javascript
let file = fs.open(name);
let buffer = new Uint8Array(5)
let p1 = file.read(buffer);
let p2 = file.read(buffer);
await Promise.all(p1, p2);
```

```javascript
let file = fs.open(name);
let buffer = new Uint8Array(5)
let p1 = file.read(buffer);
buffer[2] = 3;
console.log(buffer); // this here is not certain whether it will have teh data read from file
// or will have just the third value 3, or combination of the two
await p1;
```

in both cases this can could modify the underlying bytes array multiple times.

Arguably the observability of a change in `buffer` before awaiting the promises is the bigger conceptual problem as that basically breaks promises behaviour - where you can not observe (and consequently race) your state around promises in any other way apart from callbacks - which make the whole thing easier.

Some experimentation(probably more required, and no documentation confirms this) with deno shows that:
1. if you have sequential calls to `read` (without await) the moment you await one of the promises - its data is populated
2. awaiting a promise from a read before one htat was already done - does not change the array buffer
3. the array buffer does not change if you do not await a promise on the read.

So it seems like the deno's implementation orders calls to `read` and skips previous ones :shrug: no idea - more investigation needed, and I would like to see their rational on this. Maybe node has some more info on this as well :shrug:  

I have the feeling we have a similar issue with `seek`, but less ... dramatic


Again I am not against this staying this way for the `experimental` module, but IMO having a way to make race conditions from js code is big no-no for any stable code.

Reply:
That's very insightful and a great catch, indeed. I hadn't considered that concurrency issue.

I align with you and would rather not block the feature on the resolution of this issue, but I would consider it a blocker for a stable release of the module 👍🏻 

No apparent solutions come to my mind at the moment, and I will look at Deno's and Node's implementation to find out how they handled it.

In the meantime, I have created a [dedicated issue to keep track of it](https://github.com/grafana/k6/issues/3354) 🙇🏻

Reply:
>from js code is big no-no for any stable code.

If we evaluate this use case so important then I disagree with the following sentence.

>I align with you and would rather not block the feature on the resolution of this issue, but I would consider it a blocker for a stable release of the module 👍🏻

We can avoid blocking this specific PR but it should have some answer before we launch the experimental module. The reason is that if we evaluate this use case so important then in the eventual case that we can't fix it with the current solution then we will have to waste a bunch of stuff.

We should have at least a rough idea of what could be the solution.

Reply:
I have a potential solution in mind that I'll post in the dedicated issue 👍🏻

Reply:
This has now been addressed by #3420 🙇🏻